runtime/src/signed_extensions/check_call_allowed.rs

```diff
@@ -0,0 +1,148 @@
+use crate::{AccountId, Balances, Call, Runtime};
+use codec::{Decode, Encode};
+use common::locks::STAKING_LOCK_ID;
+use frame_support::traits::LockIdentifier;
+use frame_system::Config;
+use scale_info::TypeInfo;
+use sp_runtime::{
+    traits::{DispatchInfoOf, SignedExtension},
+    transaction_validity::{
+        InvalidTransaction, TransactionLongevity, TransactionValidity, TransactionValidityError,
+        ValidTransaction,
+    },
+};
+use sp_std::{vec, vec::Vec};
+use staking_handler::LockComparator;
+
+const BONDING_NOT_ALLOWED: u8 = 1;
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct CheckCallAllowed<T: Config + Send + Sync>(sp_std::marker::PhantomData<T>);
+
+impl CheckCallAllowed<Runtime> {
+    /// Create new `SignedExtension` to check allowed calls in transaction
+    pub fn new() -> Self {
+        Self(sp_std::marker::PhantomData)
+    }
+}
+
+impl sp_std::fmt::Debug for CheckCallAllowed<Runtime> {
+    #[cfg(feature = "std")]
+    fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+        write!(f, "CheckCallAllowed")
+    }
+
+    #[cfg(not(feature = "std"))]
+    fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+        Ok(())
+    }
+}
+
+impl Default for CheckCallAllowed<Runtime> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl CheckCallAllowed<Runtime> {
+    // Checks to see if the account has any rivalrous locks that would conflict
+    // with the FRAME staking pallet lock
+    fn has_no_conflicting_locks(who: &AccountId) -> bool {
+        let existing_locks = Balances::locks(who);
+
+        if existing_locks.len() == 0 {
+            return true;
+        }
+
+        if existing_locks.len() == 1 && existing_locks[0].id == STAKING_LOCK_ID {
```

Comment:
You are correct. The only reason I added this and the simpler prior check is to make this function as fast as possible. So to avoid need to always call `Runtime::are_locks_conflicting`. But it is also to not consider it an invalid transaction when account is already bonded, so user will get a proper Dispatch Error detailing reason why bond() call fails.